src/platform/linux/x11/mod.rs

```diff
@@ -523,6 +535,39 @@ impl EventsLoop {
     }
 }
 
+impl EventsLoopProxy {
+    pub fn wakeup(&self) -> Result<(), EventsLoopClosed> {
+        // Update the `EventsLoop`'s `pending_wakeup` flag.
+        let display = match (self.pending_wakeup.upgrade(), self.display.upgrade()) {
+            (Some(wakeup), Some(display)) => {
+                wakeup.store(true, atomic::Ordering::Relaxed);
+                display
+            },
+            _ => return Err(EventsLoopClosed),
+        };
+
+        // Push an event on the X event queue so that methods like run_forever will advance.
+        let mut xev = ffi::XClientMessageEvent {
+            type_: ffi::ClientMessage,
+            window: self.root,
+            format: 32,
+            message_type: 0,
+            serial: 0,
+            send_event: 0,
+            display: display.display,
+            data: unsafe { mem::zeroed() },
+        };
+
+        unsafe {
+            (display.xlib.XSendEvent)(display.display, self.root, 0, 0, mem::transmute(&mut xev));
```

Comment:
OK I just tested this on X11 and you are right, this does not seem to cause `xlib.XNextEvent` to stop blocking at all (in the proxy.rs example). That said, I can't help but wonder if this code ever worked for the old `WindowProxy` code either.

I'm not very familiar with the X11 backend but will see if I can get to the bottom of this. In the meantime, do you have any ideas on why this might not be working?